File: f#/src/SodiumFRP.FSharp/_Root/Listener.fs
namespace SodiumFRP.FSharp

open System
open System.Collections.Generic

type IListenerWithWeakReference =
    abstract member Unlisten : unit -> unit

type IListener =
    abstract member Unlisten : unit -> unit
    abstract member GetListenerWithWeakReference : unit -> IListenerWithWeakReference

type IStrongListener =
    inherit IListener
    inherit IDisposable

type IWeakListener =
    inherit IListener

module Listener =

    let unlistenWeak (listener : IListenerWithWeakReference) = listener.Unlisten ()
    let unlisten (listener : IListener) = listener.Unlisten ()
    let getListenerWithWeakReference (listener : IListener) = listener.GetListenerWithWeakReference ()

    type private ActionListener(unlisten : unit -> unit) =
        interface IListener with
            member __.Unlisten () = unlisten ()
            member this.GetListenerWithWeakReference () = upcast this
        interface IListenerWithWeakReference with
            member __.Unlisten () = unlisten ()
        interface IDisposable with
            member __.Dispose () = unlisten ()

    type private CompositeListener<'T when 'T :> IListener>(listeners : IReadOnlyList<'T>) =
        let listenersWeak = lazy (Seq.map getListenerWithWeakReference listeners |> Array.ofSeq)
        interface IListener with
            member __.Unlisten () = for listener in listeners do listener.Unlisten ()
            member this.GetListenerWithWeakReference () = upcast this
        interface IListenerWithWeakReference with
            member __.Unlisten () = for listener in listenersWeak.Value do listener.Unlisten ()
    
    type private CompositeWeakListener(listeners) =
        inherit CompositeListener<IWeakListener>(listeners)
        interface IWeakListener with
            member this.Unlisten () = unlisten this
    
    type private CompositeStrongListener(listeners) =
        inherit CompositeListener<IStrongListener>(listeners)
        interface IStrongListener with
            member this.Unlisten () = unlisten this
            member this.Dispose () = unlisten this

    let internal fromAction a : IListener = upcast new ActionListener(a)
    let internal fromNodeAndTarget (node : Node<_>) target = fromAction (fun () -> node.Unlink target)
    let empty = fromAction(fun() -> ())

    let fromList listeners : IListener = upcast new CompositeListener<#IListener>(listeners)
    let fromWeakList listeners : IWeakListener = upcast new CompositeWeakListener(listeners)
    let fromStrongList listeners : IStrongListener = upcast new CompositeStrongListener(listeners)
    let fromSeq listeners = List.ofSeq listeners |> fromList

    let append (l1 : IListener) (l2 : IListener) : IListener = upcast new ActionListener(fun() ->
        l1.Unlisten()
        l2.Unlisten())

type internal IKeepListenersAlive =
    abstract member KeepListenerAlive : IListener -> unit
    abstract member StopKeepingListenerAlive : IListener -> unit
    abstract member Use : IKeepListenersAlive -> unit

File: dotnet/src/Sodium/Sodium.Core.Frp/Listener.cs
using System;
using System.Collections.Generic;
using System.Linq;

namespace Sodium.Frp
{
    /// <summary>
    ///     An interface representing an stream event listener which does not keep the stream from being garbage collected.
    ///     This may be used to stop listening on a stream by calling <see cref="Unlisten" />.
    /// </summary>
    public interface IListenerWithWeakReference
    {
        void Unlisten();
    }

    /// <summary>
    ///     An interface representing an stream event listener.  This may be used to stop listening on a stream by calling
    ///     <see cref="Unlisten" />.
    /// </summary>
    public interface IListener
    {
        void Unlisten();

        IListenerWithWeakReference GetListenerWithWeakReference();
    }

    /// <summary>
    ///     An interface representing an stream event listener.  This may be used to stop listening on a stream by calling
    ///     <see cref="IListener.Unlisten" />.
    /// </summary>
    public interface IStrongListener : IListener, IDisposable
    {
    }

    /// <summary>
    ///     An interface representing an stream event listener.  This may be used to stop listening on a stream by calling
    ///     <see cref="IListener.Unlisten" />.
    /// </summary>
    public interface IWeakListener : IListener
    {
    }

    internal interface IKeepListenersAlive
    {
        void KeepListenerAlive(IListener listener);
        void StopKeepingListenerAlive(IListener listener);
        void Use(IKeepListenersAlive childKeepListenersAlive);
    }

    internal static class ListenerInternal
    {
        internal static readonly IListener EmptyImpl = EmptyListener.Instance;
        internal static readonly IWeakListener EmptyWeakImpl = EmptyListener.Instance;
        internal static readonly IStrongListener EmptyStrongImpl = EmptyListener.Instance;

        internal static IListener CreateFromNodeAndTarget<T>(Node<T> node, Node<T>.Target target) =>
            new ActionListener(() => node.Unlink(target));

        internal static IListener CreateFromAction(Action unlisten) => new ActionListener(unlisten);

        internal static IListener CreateCompositeImpl<T>(IReadOnlyList<T> listeners) where T : IListener =>
            new CompositeListener<T>(listeners);

        internal static IWeakListener CreateWeakCompositeImpl(IReadOnlyList<IWeakListener> listeners) =>
            new CompositeWeakListener(listeners);

        internal static IStrongListener CreateStrongCompositeImpl(IReadOnlyList<IStrongListener> listeners) =>
            new CompositeStrongListener(listeners);

        private class EmptyListener : IStrongListener, IWeakListener, IListenerWithWeakReference
        {
            public static readonly EmptyListener Instance = new EmptyListener();

            private EmptyListener()
            {
            }

            public void Unlisten()
            {
            }

            public IListenerWithWeakReference GetListenerWithWeakReference() => this;

            public void Dispose()
            {
            }
        }

        /// <summary>
        ///     A listener which runs the specified action when it is disposed.
        /// </summary>
        private class ActionListener : IListener, IListenerWithWeakReference
        {
            private readonly Action unlisten;

            /// <summary>
            ///     Creates a listener which runs the specified action when it is disposed.
            /// </summary>
            /// <param name="unlisten">The action to run when this listener should stop listening.</param>
            internal ActionListener(Action unlisten) => this.unlisten = unlisten;

            public void Unlisten() => this.unlisten();
            public IListenerWithWeakReference GetListenerWithWeakReference() => this;
        }

        private class CompositeListener<T> : IListener, IListenerWithWeakReference
            where T : IListener
        {
            private readonly IReadOnlyList<T> listeners;

            public CompositeListener(IReadOnlyList<T> listeners) => this.listeners = listeners;

            public void Unlisten()
            {
                foreach (T l in this.listeners)
                {
                    l?.Unlisten();
                }
            }

            public IListenerWithWeakReference GetListenerWithWeakReference()
            {
                return new CompositeWeakListener(
                    this.listeners.Select(l => l.GetListenerWithWeakReference()).ToArray());
            }

            private class CompositeWeakListener : IListenerWithWeakReference
            {
                private readonly IReadOnlyList<IListenerWithWeakReference> weakListeners;

                public CompositeWeakListener(IReadOnlyList<IListenerWithWeakReference> weakListeners) =>
                    this.weakListeners = weakListeners;

                public void Unlisten()
                {
                    foreach (IListenerWithWeakReference l in this.weakListeners)
                    {
                        l?.Unlisten();
                    }
                }
            }
        }

        private class CompositeStrongListener : CompositeListener<IStrongListener>, IStrongListener
        {
            public CompositeStrongListener(IReadOnlyList<IStrongListener> listeners)
                : base(listeners)
            {
            }

            public void Dispose()
            {
                this.Unlisten();
            }
        }

        private class CompositeWeakListener : CompositeListener<IWeakListener>, IWeakListener
        {
            public CompositeWeakListener(IReadOnlyList<IWeakListener> listeners)
                : base(listeners)
            {
            }
        }
    }
}